### ISSUE TITLE
Inline footnotes?

### ISSUE THREAD

**Issue by rauschma** [jgm/djot#286]
I like [Pandoc’s inline footnotes](https://pandoc.org/MANUAL.html#extension-inline_notes):

```
Here is an inline note.^[Inline notes are easier to write, since
you don't have to pick an identifier and move down to type the
note.]
```

Are there any plans to support them in djot? Would it use the same syntax?

**Comment by denismaier** [comment]
I also think inline footnotes are great feature that should be supported in djot as well.

**Comment by Omikhleia** [comment]
The syntax `^[...` is however somewhat ambiguous, with `^...^` currently used for super-scripting, and the possibly in the later case to have a span as content, e.g. `^[something super-scripted]{.colored}^. I think Djot tries to avoid such ambiguities, and the look-ahead then necessary to resolve them. So if inline footnotes are eventually supported, some syntax elements may have to be changed. (?)

**Comment by bpj** [comment]
Meanwhile a Pandoc filter which turns a span with a certain class (here `.inote`) into a note is trivial:

```lua
Span = function(span)
  if span.classes:includes('inote') then
    return pandoc.Note(span.content)
  end
  return nil
end
```

N.B. It relies on the fact that `pandoc.Note` automatically wraps the list of inlines it receives in a Plain element.

**Comment by dereuromark** [comment]
I've been thinking about this for djot-php and wanted to share some thoughts on possible approaches.

## The Syntax Conflict

As @bpj noted, `^[` is problematic because:
- `^text^` is superscript
- `^[text]{.class}^` is a valid superscripted span
- Resolving `^[` requires lookahead to determine if it's superscript-starting-with-bracket or inline-footnote

## Possible Approaches

### 1. New syntax like `^[...]` (requires spec change)

Would need disambiguation rules. Perhaps `^[` only starts an inline footnote if not followed by `]^` within the same inline context? Gets complicated.

### 2. Span with class convention: `[content]{.fn}`

Works today with filters/renderers. The footnote content is the span content:

```djot
Here is text[An inline footnote with *formatting*]{.fn} that continues.
```

Pros:
- No parser changes needed
- Content supports full inline formatting
- Follows existing djot patterns

Cons:
- More verbose than `^[...]`
- Convention, not spec

### 3. Span with attribute: `[^]{note="content"}`

```djot
Here is text[^]{note="An inline footnote"} that continues.
```

Pros:
- Visually similar to reference footnotes `[^label]`
- Compact marker in flow

Cons:
- Content in attributes is awkward (escaping quotes, no formatting)
- Feels like a workaround

## Recommendation

The **span-with-class approach** (`[content]{.fn}`) seems best for current djot:

1. Works today without spec changes
2. Content is first-class (supports formatting, nesting)
3. Consistent with djot's "attributes on everything" philosophy
4. Easy to implement in any djot processor via filters

For djot-php I'm planning to support this via a render event that converts `[...]{.fn}` spans to proper footnotes. Other implementations could do the same.

If inline footnotes were to become a core djot feature, the syntax question would need more thought - but the span approach provides a good interim solution that processors can adopt today.

**Comment by AvtechScientific** [comment]
Why not just `text[^ body of a short inline footnote]`. It's consistent with, but also different from the regular case when there is no whitespace after `^`: `text[^ref]`

**Comment by vassudanagunta** [comment]
> Why not just `text[^ body of a short inline footnote]`

I really like this. We can generalize:

```
This sentence has a footnote.[^ A ref is not needed.]

This sentence has a footnote.[^ref1 The ref allows this inline footnote to be referenced elsewhere.]

This sentence has a footnote.[^ref1]

This sentence has a footnote.[^ref2]

[^ref2]: This is a non-inline footnote.
```

**Comment by dereuromark** [comment]
[^ body] is much better than ^[...] because it avoids the superscript conflict entirely.

That said, I like the simple form more than the generalized one. [^ body] is readable and easy to explain: whitespace after ^ means “this is note content, not a reference label.” But [^ref1 body] starts to overload one construct with two grammars whose meaning hinges on the first space. That is parseable, but it feels a bit less djot-like, since djot usually tries to avoid subtle whitespace-driven ambiguity.

A few points/concerns:

1. Whitespace becomes semantic.
[^ref] means reference, but [^ ref] means inline note. That’s compact, but it makes a single space change the node type.
 1. It’s visually subtle.
In source, the difference between [^ref] and [^ ref] is easy to miss, especially in longer prose or diffs.
 1. It only naturally covers “short inline content.”
Standard footnotes can hold block content. This syntax strongly suggests inline-only content, so it’s not feature-equivalent to normal footnotes.
 1. It needs very clear escaping rules.
If the inline note contains ], brackets, attributes, or nested constructs, the termination rules have to be nailed down carefully.

For the generalized [^ref body], the issues are bigger:

 1. It overloads label and content into one bracket form.
Now the parser must decide whether text after ^ is:
  - a plain reference label: [^ref]
  - an unlabeled inline note: [^ body]
  - a labeled inline note: [^ref body]
 2. The first space becomes a delimiter.
That means labels can’t freely contain spaces anymore, or at least not without changing the current interpretation model.
 3. It creates awkward edge cases.
Examples:
  - [^ref ] — labeled inline note with empty content?
  - [^ ref] — unlabeled note or label " ref"?
  - [^a b c] — label a, content b c, or label a b c?
 4. It mixes two different concepts.
A reference points to a separately defined footnote. An inline footnote is itself a definition. Combining both in one syntax is clever, but 
conceptually muddier.
 5. Reuse/precedence gets messy.
If [^ref inline body] defines a footnote, what happens if there is also a later [^ref]: block footnote? Which wins? Are duplicate inline definitions 
allowed?
 6. Referencing behavior needs rules.
If [^ref inline body] creates a reusable note, can [^ref] appear before it? After it? Multiple times? Can multiple inline definitions share a label?

So the short version is:
 - [^ body]: plausible, but whitespace-sensitive and narrower than full footnotes.
 - [^ref body]: more powerful, but much messier because it turns one syntax into three related grammars.

And: 
As such I think for now `[content]{.fn}` is still the best alternative so far for this inline topic.

   1. No whitespace ambiguity
  [...]{.fn} is always a span plus attributes. There is no [^ref] vs [^ ref] distinction.
   2. No overloaded [^ ... ] grammar
  We keep standard footnotes as they are: [^label] + [^label]: .... The .fn form is a separate convention, not a second/third meaning of the same 
  syntax.
   3. No label parsing problem
  The content is just normal inline content inside the span. There is no need to split “label” from “body” at the first space.
   4. Formatting works naturally
  Because the content is real span content, it already supports emphasis, code, links, nested inline markup, etc. We don’t need special rules for  “inline footnote body parsing.”
   5. No precedence/conflict rules with real footnotes
  There’s no question like “does [^ref body] define something reusable?” or “what if [^ref]: ... also exists?” The .fn span is just its own thing.

You can test it and see it in action in the sandbox.

**Comment by AvtechScientific** [comment]
>  That is parseable, but it feels a bit less djot-like, since djot usually tries to avoid subtle whitespace-driven ambiguity.

I find whitespace(start/end of line) being a part of a syntax a good thing that makes it lightweight, readable and fast/easy to type while eliminating  ambiguity. E.g. ` *` as opener for bold that must precede a non-space char and `* ` as closer for bold that also sticks to a non-space char. So `this *is bold*`, but `5 * 3 * 2 = 30`.
  
> * [^a b c] — label a, content b c, or label a b c?

By definition it's - label a, content b c. One can have  [^a_b_c body] if needed.

> 5. Reuse/precedence gets messy.
>    If [^ref inline body] defines a footnote, what happens if there is also a later [^ref]: block footnote? Which wins? Are duplicate inline definitions allowed?

The same question you can ask regarding regular footnotes that collide. Whatever you do there - do here.

> 6. Referencing behavior needs rules.
>    If [^ref inline body] creates a reusable note, can [^ref] appear before it? 

probably not.

> After it? Multiple times?

Sure, both. Why not?

> Can multiple inline definitions share a label?

No, as you attempt to redefine an existing label.

**Comment by jgm** [comment]
Currently, footnote references can come before the notes to which they refer. So, I don't see why it should be any different if we were to adopt this syntax.

**Comment by vassudanagunta** [comment]
We could generalize/unify even further, making this _the_ syntax for footnote declarations, a breaking change for existing footnotes:


```
[^ref2 This is a non-inline footnote
       declaration, distinguished by
       being standalone.]

[^ref3 This is a 
       multi-line 
      

       # multi-block

       footnote
       ]
```

Perhaps it goes too far?

**Comment by jgm** [comment]
Yes I think it goes too far. `[..]` in djot are a container for inlines, not blocks.
